java/arcs/android/common/resurrection/DbHelper.kt

```diff
@@ -0,0 +1,242 @@
+import android.database.sqlite.SQLiteOpenHelper
+import android.os.Parcel
+import android.os.PersistableBundle
+import androidx.annotation.VisibleForTesting
+import arcs.core.storage.StorageKey
+import arcs.core.storage.StorageKeyParser
+import arcs.core.storage.driver.RamDiskStorageKey
+import arcs.core.storage.driver.VolatileStorageKey
+import arcs.core.storage.referencemode.ReferenceModeStorageKey
+
+/**
+ * Database abstraction layer for resurrection.
+ */
+@VisibleForTesting(otherwise = VisibleForTesting.PACKAGE_PRIVATE)
+class DbHelper(
+    context: Context,
+    dbName: String = RESURRECTION_DB_NAME
+) : SQLiteOpenHelper(
+    context,
+    dbName,
+    /* cursorFactory = */ null,
+    RESURRECTION_DB_VERSION
+) {
+    init {
+        // Pre-initialize parsers.
+        // TODO: Make a master storage key parsing initializer in arcs.core.storage.
+        RamDiskStorageKey.registerParser()
+        VolatileStorageKey.registerParser()
+        ReferenceModeStorageKey.registerParser()
+    }
+
+    override fun onCreate(db: SQLiteDatabase) {
+        db.transaction { CREATE.forEach(db::execSQL) }
+    }
+
+    override fun onUpgrade(db: SQLiteDatabase, oldVersion: Int, newVersion: Int) = Unit
+
+    /**
+     * Stores a [ResurrectionRequest] in the database.
+     */
+    fun registerRequest(request: ResurrectionRequest) = writableDatabase.useTransaction {
+        val requestContent = ContentValues()
+            .apply {
+                put("component_package", request.componentName.packageName)
+                put("component_class", request.componentName.className)
+                put("component_type", request.componentType.name)
+                put("intent_action", request.intentAction)
+                val extrasBlob = if (request.intentExtras != null) {
+                    with(Parcel.obtain()) {
+                        writeTypedObject(request.intentExtras, 0)
+                        marshall()
+                    }
+                } else null
+                put("intent_extras", extrasBlob)
+            }
+        insertWithOnConflict(
+            "resurrection_requests",
+            null,
+            requestContent,
+            SQLiteDatabase.CONFLICT_REPLACE
+        )
+
+        delete(
+            "requested_notifiers",
+            "component_package = ? AND component_class = ?",
+            arrayOf(
+                request.componentName.packageName,
+                request.componentName.className
+            )
+        )
+
+        val notifierValues = ContentValues()
+        request.notifyOn.forEach {
+            notifierValues.put(
+                "component_package",
+                request.componentName.packageName
+            )
+            notifierValues.put(
+                "component_class",
+                request.componentName.className
+            )
+            notifierValues.put("notification_key", it.toString())
+            insert("requested_notifiers", null, notifierValues)
+        }
```

Comment:
I'm a bit confused by this delete and then insert. Is this just to change the notification key?

Reply:
It's to clear out any registered keys before inserting the new ones. This approach is just simpler than trying to diff what's in the database versus what's in the ResurrectionRequest, and issuing granular INSERT/DELETE statements.